x-pack/legacy/plugins/apm/public/components/app/ServiceMap/index.tsx

```diff
@@ -37,37 +54,159 @@ ${theme.euiColorLightShade}`,
   margin: `-${theme.gutterTypes.gutterLarge}`
 };
 
+const MAX_REQUESTS = 5;
+
 export function ServiceMap({ serviceName }: ServiceMapProps) {
-  const {
-    urlParams: { start, end }
-  } = useUrlParams();
+  const callApmApi = useCallApmApi();
+  const license = useLicense();
+  const { search } = useLocation();
+  const { urlParams, uiFilters } = useUrlParams();
+  const { notifications } = useApmPluginContext().core;
+  const params = useDeepObjectIdentity({
+    start: urlParams.start,
+    end: urlParams.end,
+    environment: urlParams.environment,
+    serviceName,
+    uiFilters: {
+      ...uiFilters,
+      environment: undefined
+    }
+  });
+
+  const renderedElements = useRef<ElementDefinition[]>([]);
+  const openToast = useRef<string | null>(null);
+
+  const [responses, setResponses] = useState<ServiceMapAPIResponse[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [percentageLoaded, setPercentageLoaded] = useState(0);
+  const [, _setUnusedState] = useState(false);
+
+  const elements = useMemo(() => getCytoscapeElements(responses, search), [
+    responses,
+    search
+  ]);
+
+  const forceUpdate = useCallback(() => _setUnusedState(value => !value), []);
```

Comment:
This was kind of a hack, and we can probably make it a little better by just using `setState` properly. We can also do that a little later though.

Reply:
Agree, would be nice to find a more elegant solution.